### ISSUE TITLE
Multiple GPUs: Illegal Memory Access

### ISSUE THREAD

**Issue by caandewiel** [rosinality/stylegan2-pytorch#44]
Hi, thanks for the great work. This is the same problem as https://github.com/rosinality/stylegan2-pytorch/issues/13. When initializing either the generator or discriminator on `cuda:1` while having 2+ GPUs available, an illegal memory access is triggered.

When using torch.distributed on a multi-gpu machine, this does not occur, since you manually set the cuda device to only one device, so device -1 is always that GPU. However, my CUDA skills are not great and I do not know how this can be solved, but it looks like an easy fix to me.

**Comment by rosinality** [comment]
Could you use CUDA_VISIBLE_DEVICES=1? It will be easier than use device strings.

**Comment by caandewiel** [comment]
I tried that and it works. However, for my current research I really need two GPUs so it’s not a viable solution. Do you have an idea what could be causing this issue?

**Comment by rosinality** [comment]
That means you need to locate generators at cuda:0 and discriminators at cuda:1? I think

```python
gen = gen.to('cuda:0')
z = torch.randn(1, 512, device='cuda:0')
img = gen([x])
img = img.to('cuda:1')
dis = dis.to('cuda:1')
out = dis(img)
```

should work.

**Comment by caandewiel** [comment]
Also tried that already. This doesn't work either. The same error will be thrown at a random line (different every run). It looks like the custom ops break when you have multiple visible cuda devices regardless of the device the generator is on.

**Comment by rosinality** [comment]
I got it. It will require some modifications in cuda kernels.

**Comment by rosinality** [comment]
Hmm In my cases it just works. Could you give me an error messages with CUDA_LAUNCH_BLOCKING=1?

**Comment by caandewiel** [comment]
I ran on Titan V, RTX 2080ti and GTX 1080 ti. They all yield the same error.
- Driver: 415.27
- CUDA: 10.0 (also tried 10.1 and get the same error)
```
Traceback (most recent call last):
  File "train.py", line 175, in <module>
    latent_out = g_ema.style(noise_sample)
  File "/home/cwiel/miniconda3/envs/sg2/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/miniconda3/envs/sg2/lib/python3.6/site-packages/torch/nn/modules/container.py", line 100, in forward
    input = module(input)
  File "/home/userl/miniconda3/envs/sg2/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/storage/user/stylegan2/model.py", line 213, in forward
    out = F.linear(input, self.weight * self.scale)
RuntimeError: CUDA error: an illegal memory access was encountered
```

Line 213 contains (from the `EqualLinear` class:
```
def forward(self, input):
        if self.activation:
-->         out = F.linear(input, self.weight * self.scale)
            out = fused_leaky_relu(out, self.bias * self.lr_mul)

        else:
            out = F.linear(
                input, self.weight * self.scale, bias=self.bias * self.lr_mul
            )

        return out
```

**Comment by rosinality** [comment]
Sorry, could you give me the script that I could replicate the error? It seems like that it will be difficult to trace the error using error logs.

**Comment by caandewiel** [comment]
I quickly made a really simple script causing this error for me (don't mind the encoder architecture, it's just to trigger the error :P).

```
import torch
import torchvision

from torch import nn
from model import Generator, Discriminator
from dataset import MultiResolutionDataset

def data_sampler(dataset, shuffle):
    return torch.utils.data.RandomSampler(dataset)

transform = torchvision.transforms.Compose(
    [
        torchvision.transforms.RandomHorizontalFlip(),
        torchvision.transforms.ToTensor(),
        torchvision.transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5), inplace=True),
    ]
)

dataset = MultiResolutionDataset('path/to/ffhq', transform, 256)
loader = torch.utils.data.DataLoader(
    dataset,
    batch_size=1,
    sampler=data_sampler(dataset, shuffle=True),
    drop_last=True,
)

class Encoder(nn.Module):
    def __init__(self):
        super(Encoder, self).__init__()

        self.conv_layer = nn.Sequential(
            nn.Conv2d(3, 64, 3, stride=2),
        )
        self.fc = nn.Linear(196608, 14 * 512)

    def forward(self, x):
        out = self.conv_layer(x)
        out = self.fc(x.view(x.size(0), -1))

        return out.view((-1, 14, 512))

encoder = Encoder().to('cuda:0')
generator = Generator(256, 512, 8).to('cuda:1')
discriminator = Discriminator(256).to('cuda:0')

imgs = next(iter(loader))
imgs = imgs.to('cuda:0')
w = encoder(imgs).to('cuda:1')

out, _ = generator([w], input_is_latent=True)

out = out.to('cuda:0')
discriminator(out)
```

**Comment by rosinality** [comment]
Strangely it works for me. Now I suspect maybe it is system/device or cudnn or pytorch related problems.

**Comment by caandewiel** [comment]
Which versions of pytorch/cudnn are you using?

**Comment by rosinality** [comment]
I have used pytorch 1.3.1.

**Comment by caandewiel** [comment]
Guess I found the issue, doesn't seem to be related to your code. Switching all `cuda:0` to `cuda:1` and vice versa seems to fix the problem after downgrading to PyTorch 1.3.1. The workstationm I am using has a Titan V added later, so in the second slot, making it GPU:1, while PyTorch sees it as GPU:0 since it has more memory than the default RTX 2080ti.

**Comment by LiUzHiAn** [comment]
Sorry, I get into this issue. And the followings are my log info with the runnig script `CUDA_LAUNCH_BLOCKING=1 CXX=g++ python3 -m torch.distributed.launch --nproc_per_node=8 --master_port=6250 train.py --batch 4 --size 1024 --iter 150000 <MyLmdbDataPath> `

```
Traceback (most recent call last):

File "train.py", line 532, in <module>

train(args, loader, generator, discriminator, g_optim, d_optim, g_ema, device)

File "train.py", line 308, in train

sample, _ = g_ema([sample_z])

File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl

return forward_call(*input, **kwargs)

File "/XXX/XXX/XXX/trainer/model.py", line 558, in forward

out = conv1(out, latent[:, i], noise=noise1)

File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl

return forward_call(*input, **kwargs)

File "/XXX/XXX/XXX/trainer/model.py", line 362, in forward

out = self.noise(out, noise=noise)

File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl

return forward_call(*input, **kwargs)

File "/XXX/XXX/XXX/trainer/model.py", line 314, in forward

noise = image.new_empty(batch, 1, height, width).normal_() 

RuntimeError: CUDA error: an illegal memory access was encountered

terminate called after throwing an instance of 'c10::CUDAError'

what(): CUDA error: an illegal memory access was encountered

Exception raised from create_event_internal at /pytorch/c10/cuda/CUDACachingAllocator.cpp:1055 (most recent call first):

frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x42 (0x7f18db3a4a22 in /opt/conda/lib/python3.8/site-packages/torch/lib/libc10.so)

frame #1: <unknown function> + 0x10aa3 (0x7f18db605aa3 in /opt/conda/lib/python3.8/site-packages/torch/lib/libc10_cuda.so)

frame #2: c10::cuda::CUDACachingAllocator::raw_delete(void*) + 0x1a7 (0x7f18db607147 in /opt/conda/lib/python3.8/site-packages/torch/lib/libc10_cuda.so)

frame #3: c10::TensorImpl::release_resources() + 0x54 (0x7f18db38e5a4 in /opt/conda/lib/python3.8/site-packages/torch/lib/libc10.so)

frame #4: <unknown function> + 0xa2822a (0x7f19801ac22a in /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_python.so)

frame #5: <unknown function> + 0xa282c1 (0x7f19801ac2c1 in /opt/conda/lib/python3.8/site-packages/torch/lib/libtorch_python.so)

frame #6: <unknown function> + 0x1932c6 (0x56200475d2c6 in /opt/conda/bin/python3)

frame #7: <unknown function> + 0x15878b (0x56200472278b in /opt/conda/bin/python3)

frame #8: <unknown function> + 0x1592ac (0x5620047232ac in /opt/conda/bin/python3)

frame #9: <unknown function> + 0x158e77 (0x562004722e77 in /opt/conda/bin/python3)

frame #10: <unknown function> + 0x158e60 (0x562004722e60 in /opt/conda/bin/python3)

frame #11: <unknown function> + 0x176057 (0x562004740057 in /opt/conda/bin/python3)

frame #12: PyDict_SetItemString + 0x61 (0x5620047613c1 in /opt/conda/bin/python3)

frame #13: PyImport_Cleanup + 0x9d (0x56200479faad in /opt/conda/bin/python3)

frame #14: Py_FinalizeEx + 0x79 (0x5620047d1a49 in /opt/conda/bin/python3)

frame #15: Py_RunMain + 0x183 (0x5620047d3893 in /opt/conda/bin/python3)

frame #16: Py_BytesMain + 0x39 (0x5620047d3ca9 in /opt/conda/bin/python3)

frame #17: __libc_start_main + 0xf3 (0x7f1987499f93 in /usr/lib64/libc.so.6)

frame #18: <unknown function> + 0x1e21c7 (0x5620047ac1c7 in /opt/conda/bin/python3)

ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: -6) local_rank: 0 (pid: 255) of binary: /opt/conda/bin/python3

ERROR:torch.distributed.elastic.agent.server.local_elastic_agent:[default] Worker group failed

INFO:torch.distributed.elastic.agent.server.api:[default] Worker group FAILED. 3/3 attempts left; will restart worker group

INFO:torch.distributed.elastic.agent.server.api:[default] Stopping worker group

INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous'ing worker group

INFO:torch.distributed.elastic.agent.server.api:[default] Rendezvous complete for workers. Result:
```

My env is PyTorch 1.9.0 + cuda11.1, and I have added `1.9.1` to support conv2d_gradfix in the high version as discussed in #252. Could you help to give some advice to fix it? Will this be due to the Pytorch version problem?

**Comment by LiUzHiAn** [comment]
Please consider reducing the `--n_sample` arg, which is set to be 64 by default. This may cause OOM problem, though not very related to the illegal memory access.
